Positive Definite manifold distance

sylvchev opened the pull request:
Following PR #70, this is the PR part which focus on the correction of the distance for positive definite manifold. The part focusing on the update of Travis file is in PR #72.

NicolasBoumal commented:
Hello Sylvain,

Thanks for the separated PR. I looked at your test code for the distance. I suppose you also checked that the new code for dist is consistent with log, in the sense that M.dist(x, y) is the same as M.norm(x, M.log(x, y)). Can you please confirm this? As soon as you do, I'll just go ahead and merge this.

Thanks again!
Nicolas

sylvchev commented:
Thank you, Nicolas. The new distance is consistent with the log. I encountered some numerical  stability problem with the code of `norm` and the `np.linalg.solve`. It took me some time to figure it out, but it is already mentioned in the code of `exp`.

I updated the unit tests with this verification and I added some others.

NicolasBoumal commented on pymanopt/manifolds/psd.py:
```diff
@@ -79,7 +79,9 @@ def retr(self, X, G):
     def norm(self, x, u):
-        return la.norm(la.solve(x, u))
+        c = np.linalg.cholesky(x)
+        c_inv = np.linalg.inv(c)
```
It's usually bad practice to explicitly invert a matrix. Since you only need to apply the inverse to certain matrices, it should be possible to call a linear system solver directly (and hopefully you can tell the solver that c is triangular, if it indeed is -- I'm not familiar with np.linalg).

sylvchev replied:
You are absolutely right. I did some tests with `scipy.linalg.solve_triangular`, which is (in that case) as fast as cholesky+inv, but I ran into the same problem as the regular np.linalg.solve : the result is not reliable as the results are not always positive definite. I also tried `scipy.linalg.sqrtm` and the computation of c_inv via eigenvalue decomposition, which are both slower than cholesky+inv.

NicolasBoumal replied:
Understood, thanks for these tests -- can you add a short comment in the code to that effect? This way, other people who wonder about this will be informed too. Thanks!

sylvchev commented:
I just added some unit tests and ensure coherent notation for numpy/scipy in the PositiveDefinite package.

NicolasBoumal commented on pymanopt/manifolds/psd.py:
```diff
@@ -79,6 +79,9 @@ def retr(self, X, G):
 
     def norm(self, x, u):
+        # This implementation is as fast as np.linalg.solve_triangular and is
+        # more stable, as the solver tends to output non positive definite
```
as the solver --> as the above

(to avoid confusion -- I'd change it myself, but didn't find the option on github interface, sorry :/)

NicolasBoumal commented:
Excellent, thanks a lot Sylvain!

sweichwald commented:
Thanks @sylvchev and @NicolasBoumal !
